### PR TITLE
Modify documentation to make it more accessible to beginners

### PR DESCRIPTION
**Pull request by linuslagerhjelm**
Changed the code examples in order to make them self explanatory to those who are new to jwt. Also tried to clarify inexplicit parts of the description of a jwt.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,9 +8,9 @@ A [go](http://www.golang.org) (or 'golang' for search engine friendliness) imple
 
 In short, it's a signed JSON object that does something useful (for example, authentication).  It's commonly used for `Bearer` tokens in Oauth 2.  A token is made of three parts, separated by `.`'s.  The first two parts are JSON objects, that have been [base64url](http://tools.ietf.org/html/rfc4648) encoded.  The last part is the signature, encoded the same way.
 
-The first part is called the header.  It contains the necessary information for verifying the last part, the signature.  For example, which encryption method was used for signing and what key was used.
+The first part is called the header.  It contains the necessary information for verifying the last part, the signature. A token header *typically* consists of two parts: the type of the token, which is JWT, and the hashing algorithm such as HMAC SHA256 or RSA. [(jwt.io)](https://jwt.io/introduction/)
```
Comment (dgrijalva) [inline README.md]: 'at least two fields'.  Many types of tokens have more than two fields.

Comment (dgrijalva) [inline README.md]: Why is there a link to jwt.io?  Maybe some context around it?

File: README.md [hunk 2]
```diff
@@ -23,10 +23,10 @@ Parsing and verifying tokens is pretty straight forward.  You pass in the token
 ```go
 	token, err := jwt.Parse(myToken, func(token *jwt.Token) (interface{}, error) {
 		// Don't forget to validate the alg is what you expect:
-		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
 			return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
 		}
-		return myLookupKey(token.Header["kid"]), nil
+		return []byte(mySigningKey), nil
```
Comment (dgrijalva) [inline README.md]: Why are we simplifying away a common use case?

File: README.md [hunk 3]
```diff
@@ -45,6 +45,7 @@ Parsing and verifying tokens is pretty straight forward.  You pass in the token
 	token.Claims["foo"] = "bar"
 	token.Claims["exp"] = time.Now().Add(time.Hour * 72).Unix()
 	// Sign and get the complete encoded token as a string
+  // mySigningKey could for example be any random string
```
Comment (dgrijalva) [inline README.md]: * This is only true if you're using a symmetric signing method
* It's not a string, it's a `[]byte`
* 'any random string' is not that helpful as guidance

**Comment by dgrijalva** [comment]
All of these changes make the assumption that everyone uses the HMAC signing method.  There are many use cases for this library that are made less clear by this assumption in the documentation.

File: README.md (hunk 2 above)
Comment (linuslagerhjelm) [inline README.md]: This might very well be a common use case. I do believe, however, that it is not that clear what "myLookupKey" should be.

**Comment by linuslagerhjelm** [comment]
I believe that the original example was incomplete because, for example, tokens created by the example code were rendered invalid by the code in the parsing example since they did not use the same signing method. I would argue that it would be more helpful to beginners if they could have a working copy-paste example they could start from. Whereas more advanced users could derive other use cases from the documentation. Another possibility would be to provide additional (separate) examples to demonstrate other important use cases.

**Comment by techjanitor** [comment]
I think the existing example is terrible because it's not clear what the 'myLookupKey' function is supposed to do. How is it expected that people who aren't intimately familiar with JWT to ascertain what is going on? I've been using this library for like a year, directly returning a byte, and I just barely figured out what the example/kid claim means and how you're supposed to use it.

**Comment by dgrijalva** [comment]
@techjanitor did you read the documentation? https://godoc.org/github.com/dgrijalva/jwt-go#Keyfunc
Did you find that unhelpful?

@linuslagerhjelm I think it's a valid point that the write and read operations in the example should be compatible.  It's probably worth moving the examples out of the README and into separate files where they can be automatically tested.  I don't, however, agree that the goal should be to enable copy/paste coding.

When implementing any security feature, it's important to first understand how it works, how it was intended to be used, and what it's limitations are.  There are plenty of resources for this, including jwt.io and the spec itself.  The example code in this library intends to demonstrate how this library interacts with the spec.

**Comment by techjanitor** [comment]
Yes it would be a tragedy if the examples were useful and didn't have this mysterious 'myLookupKey' function that is literally explained nowhere.

**Comment by dgrijalva** [comment]
@linuslagerhjelm what do you think of #128 as a solution to this?  I've been wanting to move example code out of the README for a while as it regularly gets out of date.

**Comment by dgrijalva** [comment]
@linuslagerhjelm I pulled most of your changes into this.  A few of the README changes were not technically accurate, so I left them out.  I also moved the documentation about HMAC keys into the hmac examples, since that stuff only applies to that key type.

**Comment by dgrijalva** [comment]
Replaced with #128.  If you feel this doesn't solve your concern, please re-open this ticket, an issue, or a new PR.

**Comment by moberemk** [comment]
For the record @dgrijalva I agree with @linuslagerhjelm and @techjanitor on `myLookupKey` being a problem--I pored over the documentation and this pull request was the **first obvious place** that explained what that function could possibly mean. It's frightfully unintuitive given Go's shared syntax for function calls and type casting.

**Comment by dgrijalva** [comment]
@moberemk fair.  What do you think of the updated example code (#128)?  The sample code in the README has been problematic forever as it seems to always get out of sync with the implementation.

**Comment by moberemk** [comment]
The updated code is better, but still not great--the `Example` functions are never actually called with an example of what a lookup function might look like. That and the lack of any kind of example data makes this example runnable only in that it compiles, not in that it can actually be run and examined to understand practical usage.

**Comment by dgrijalva** [comment]
Good point.

-dave

**Pull request closed without merging**